### PR TITLE
zc.latest.0.ticket377.util-cpp-unit-tests.0

### PR DESCRIPTION
**Pull request by defuse**
https://github.com/Electric-Coin-Company/zerocashd/issues/377

### CODE REVIEW

**Comment by defuse** [comment]
Some of the SHA256 tests fail. I don't know why even after spending a bit of time trying to figure it out. I propose we accept this then open a new ticket to find out why, so that we get the benefits of the other tests that are passing ASAP.

File: Makefile [hunk 1]
```diff
@@ -1,5 +1,5 @@
 OPTFLAGS = -march=native -mtune=native -O2
-CXXFLAGS += -g -Wall -Wextra -Werror -Wfatal-errors -Wno-unused-parameter -std=c++11 -fPIC -Wno-unused-variable
+CXXFLAGS += -g -Wall -Wextra -Wno-unused-parameter -std=c++11 -fPIC -Wno-unused-variable
```
Comment (nathan-at-least) [inline Makefile:2]: [does not block] In the long run I'm against removing `-Werror`, but in the short term, let's leave this removed, and then have a followup task to disallow any compiler warnings. Thoughts?

File: libzerocash/utils/util.cpp [hunk 2]
```diff
@@ -182,7 +182,7 @@ void concatenateVectors(const std::vector<unsigned char>& A, const std::vector<u
     result.insert(result.end(), C.begin(), C.end());
 }
 
-void sha256(unsigned char* input, unsigned char* hash, int len) {
+void sha256(const unsigned char* input, unsigned char* hash, int len) {
```
Comment (nathan-at-least) [inline libzerocash/utils/util.cpp:185]: Great! Let's start ratcheting down compiler checks whenever we can correctly introduce tighter static constraints / types.

File: tests/utilTest.cpp [hunk 3]
```diff
@@ -0,0 +1,421 @@
...
+
+    memset(bytes1, 0, 32);
+    memset(bytes2, 0, 32);
+
+    libzerocash::getRandBytes(bytes1, 32);
+    libzerocash::getRandBytes(bytes2, 32);
+
+    BOOST_CHECK( memcmp(bytes1, bytes2, 32) != 0 );
+    BOOST_CHECK( memcmp(bytes1, bytes1+16, 16) != 0 );
+}
```
Comment (nathan-at-least) [inline tests/utilTest.cpp:30]: LGTM.

**Comment by nathan-at-least** [comment]
Reviewing now.

File: tests/utilTest.cpp [hunk 4]
```diff
@@ -0,0 +1,421 @@
...
+    };
+
+    BOOST_CHECK(v1 == v2);
+
+    std::vector<bool> unevensize(4);
+    unsigned char abyte[1] = { 0x55 };
+    libzerocash::convertBytesToVector(abyte, unevensize);
+
+    v2 = { 0, 1, 0, 1 };
+    BOOST_CHECK(unevensize == v2);
```
Comment (nathan-at-least) [inline tests/utilTest.cpp:59]: Wait a minute… If the output vector is not large enough to represent the input, then the result is truncation?

This test LGTM for verifying the existing semantics, but IME those kinds of semantics are dangerous by leading to silent errors.  (Imagine, for example, a signature nonce being truncated in an unexpected way. If the attacker can use a signing service as an oracle, they may be able to reduce nonce size then probabilistically collide nonces to recover a private key, which would be disasterous.)

So, in the long run let's change this API to fail with a loud error if the output isn't large enough to hold the input.

File: tests/utilTest.cpp [hunk 5]
```diff
@@ -0,0 +1,421 @@
...
+        0, 0, 0, 1, 0, 0, 1, 0,
+        // 0xFF
+        1, 1, 1, 1, 1, 1, 1, 1
+    };
+    unsigned char output[5];
+    libzerocash::convertVectorToBytes(v, output);
+    BOOST_CHECK( memcmp(bytes, output, sizeof(bytes)) == 0 );
+
+    unsigned char onebyte[1];
+    std::vector<bool> unevensize { 1, 1, 1, 1, 1, 1, 1 };
```
Comment (nathan-at-least) [inline tests/utilTest.cpp:83]: [does not block] For test values like this, I like to increment or vary the elements so it's easier to count visually. ;-)

File: tests/utilTest.cpp [hunk 6]
```diff
@@ -0,0 +1,421 @@
...
+    libzerocash::convertVectorToBytes(unevensize, onebyte);
+    BOOST_CHECK(onebyte[0] == 127);
```
Comment (nathan-at-least) [inline tests/utilTest.cpp]: [does not block] Test LGTM.

Library code:

Hmm... Thinking about my above comment about mis-sized outputs, this presents a challenge to my proposal above.  Do we need to support non-byte-multiple inputs ever for this functions usage?  If not, then let's make it an error if the sizes do not align.

Otherwise, we may want to have two functions, one simply requires input length to be a multiple of 8, then calls the other one.  Then, only in the fewer cases where we explicitly need non-byte-rounded-sized inputs, we explicitly call the function without that check (and it should have an obvious name).

File: tests/utilTest.cpp [hunk 7]
```diff
@@ -0,0 +1,421 @@
...
+    unsigned char bytes[16];
+    for (int i = 0; i < 16; i++) {
+        bytes[i] = i;
+    }
+    std::vector<unsigned char> v(16);
+    libzerocash::convertBytesToBytesVector(bytes, v);
+    for (int i = 0; i < 16; i++) {
+        BOOST_CHECK(v.at(i) == bytes[i]);
+    }
+}
```
Comment (nathan-at-least) [inline tests/utilTest.cpp:98]: Test LGTM.

But, isn't this a standard constructor for `std::vector`?!

File: tests/utilTest.cpp [hunk 8]
```diff
@@ -0,0 +1,421 @@
...
+    for (int i = 0; i < 16; i++) {
+        v[i] = i;
+    }
+    unsigned char bytes[16];
+    memset(bytes, 0, 16);
+    libzerocash::convertBytesVectorToBytes(v, bytes);
+    for (int i = 0; i < 16; i++) {
+        BOOST_CHECK(bytes[i] == v.at(i));
+    }
+}
```
Comment (nathan-at-least) [inline tests/utilTest.cpp:111]: Test LGTM, but again, this seems like it would be in the `std::vector` API.

File: tests/utilTest.cpp [hunk 9]
```diff
@@ -0,0 +1,421 @@
...
+        0, 0, 0, 0, 0, 0, 1, 1,
+        // 0x12
+        0, 0, 0, 1, 0, 0, 1, 0,
+        // 0xFF
+        1, 1, 1, 1, 1, 1, 1, 1
+    };
+    std::vector<bool> actual_bits;
+    libzerocash::convertBytesVectorToVector(bytes, actual_bits);
+    BOOST_CHECK(actual_bits == expected_bits);
+}
```
Comment (nathan-at-least) [inline tests/utilTest.cpp:130]: Test LGTM, but again, I'm hoping we can replace a lot of this code with `std::vector`. Perhaps I'm naive about `std::vector` or C++ idioms. Here I'm referring to `std::vector<T>` ↔ `T[]` conversions (not the bit-vector creation itself).

File: tests/utilTest.cpp [hunk 10]
```diff
@@ -0,0 +1,421 @@
...
+        0, 0, 0, 0, 0, 0, 1, 1,
+        // 0x12
+        0, 0, 0, 1, 0, 0, 1, 0,
+        // 0xFF
+        1, 1, 1, 1, 1, 1, 1, 1
+    };
+    std::vector<unsigned char> actual_bytes;
+    libzerocash::convertVectorToBytesVector(bits, actual_bytes);
+    BOOST_CHECK(actual_bytes == expected_bytes);
+}
```
Comment (nathan-at-least) [inline tests/utilTest.cpp:150]: LGTM.

File: tests/utilTest.cpp [hunk 11]
```diff
@@ -0,0 +1,421 @@
...
+
+BOOST_AUTO_TEST_CASE( testConvertIntToBytesVector ) {
+    uint64_t val;
+    std::vector<unsigned char> expected;
+    std::vector<unsigned char> bytes(8);
+
+    val = 0ULL;
```
Comment (nathan-at-least) [inline tests/utilTest.cpp:157]: [does not block] Here I'm not certain about the int types.  Is `ULL` always 8 bytes, or always 8 bytes on a given platform regardless of whether `-m32` or `-m64` is passed to GCC, or 8 bytes on a given platform when one of those options is given?

If those type sizes vary, then I _think_ these test would detect the problem, so I don't think this should block, and if we see test failures on new platforms we can then address those platforms.

(One fear is if these tests do not fail correctly should `ULL` size vary, but I think they will if `ULL` were ever 4 bytes. Not sure about larger than 8 bytes, but I don't think those exist yet.)

File: tests/utilTest.cpp [hunk 12]
```diff
@@ -0,0 +1,421 @@
...
+    BOOST_CHECK( expected == bytes );
+
+    // The following two tests aren't necessarily desired results. They merely
+    // document the behavior so that we'll be alerted if it changes in the
+    // future.
+
+    val = 0xffffffffdeadbeefULL; // truncation
+    expected = { 0xde, 0xad, 0xbe, 0xef };
+    std::vector<unsigned char> small_bytes(4);
+    libzerocash::convertIntToBytesVector(val, small_bytes);
```
Comment (nathan-at-least) [inline tests/utilTest.cpp:184]: Test LGTM, but as I said above, "silent" truncation is a slippery slope to security vulnerabilities.

File: tests/utilTest.cpp [hunk 13]
```diff
@@ -0,0 +1,421 @@
...
+    BOOST_CHECK( expected == small_bytes );
+
+    val = 0xf1f2f3f401020304ULL; // bytes buffer is too big
+    // The first byte is 4 because `>> 64` is undefined, and that's the result
+    // it has on my system (note that it's the same as the original LSB).
```
Comment (nathan-at-least) [inline tests/utilTest.cpp:189]: This is a good example of why having the API be strict on the input sizes would be safer. We're now testing "undefined behavior" which changes with compilers, compiler flags, platforms, etc...

Still, if it fails noisily in the future, we can address the issue at that time.

Also, I'm happy to see the tests are trying out these edge cases!

File: tests/utilTest.cpp [hunk 14]
```diff
@@ -0,0 +1,421 @@
...
+    val = libzerocash::convertBytesVectorToInt(bytes);
+    BOOST_CHECK( expected == val );
+}
+
+BOOST_AUTO_TEST_CASE( testConcatenateTwoBoolVectors ) {
+    std::vector<bool> front = { 0, 1, 0 };
+    std::vector<bool> back = { 1, 0, 1 };
+    std::vector<bool> expected = { 0, 1, 0, 1, 0, 1 };
+    std::vector<bool> actual;
+    libzerocash::concatenateVectors(front, back, actual);
```
Comment (nathan-at-least) [inline tests/utilTest.cpp:232]: Test LGTM.

hrm...  It looks like this API does allocation into the output.  If we're going to have the library do allocation, then why not return a newly allocated output vector?  If we're not going to do allocation in the library, then let's be consistent about it.  If some functions allocate and some do not, let's clarify that in their names and/or type signatures.

File: tests/utilTest.cpp [hunk 15]
```diff
@@ -0,0 +1,421 @@
...
+}
+
+BOOST_AUTO_TEST_CASE( testConcatenateTwoByteVectors ) {
+    std::vector<unsigned char> front = { 0, 1, 0 };
+    std::vector<unsigned char> back = { 1, 0, 1 };
+    std::vector<unsigned char> expected = { 0, 1, 0, 1, 0, 1 };
+    std::vector<unsigned char> actual;
+    libzerocash::concatenateVectors(front, back, actual);
+    BOOST_CHECK( expected == actual );
+}
```
Comment (nathan-at-least) [inline tests/utilTest.cpp:243]: LGTM. One improvement, IMO, is to do `{ 0, 1, 2 }` and `{ 3, 4, 5 }` so that it's easier to read quickly.

File: tests/utilTest.cpp [hunk 16]
```diff
@@ -0,0 +1,421 @@
...
+    std::vector<unsigned char> actual;
+    libzerocash::concatenateVectors(front, middle, back, actual);
+    BOOST_CHECK( expected == actual );
+}
+
+BOOST_AUTO_TEST_CASE( testSHA256TestVectors ) {
+    unsigned char actual_hash[32];
+    libzerocash::sha256(sha256_preimage, actual_hash, 3);
+    BOOST_CHECK( memcmp(sha256_hash, actual_hash, 32) );
+}
```
Comment (nathan-at-least) [inline tests/utilTest.cpp:269]: If these are failing and the test vectors were generated with full `sha256`, the reason may be because `libzerocash::sha256` is _only_ the compression function I beileve.

If I'm correct, then let's rename it to `libzerocash::sha256_compress` to make this clear. As discussed in a mtg, let's fix `sha256` test failures in a followup ticket.

File: tests/utilTest.cpp [hunk 17]
```diff
@@ -0,0 +1,421 @@
...
+
+BOOST_AUTO_TEST_CASE( testSHA256TestVectorsCTX ) {
+    unsigned char actual_hash[32];
+    SHA256_CTX_mod ctx256;
+    libzerocash::sha256(&ctx256, sha256_preimage, actual_hash, 3);
+    BOOST_CHECK( memcmp(sha256_hash, actual_hash, 32) );
+}
```
Comment (nathan-at-least) [inline tests/utilTest.cpp:297]: LGTM.

File: tests/utilTest.cpp [hunk 18]
```diff
@@ -0,0 +1,421 @@
...
+BOOST_AUTO_TEST_CASE( testHashBoolVectorToBoolVectorCTX ) {
+    SHA256_CTX_mod ctx256;
+
+    std::vector<bool> preimage(SHA256_PREIMAGE_BYTES * 8);
+    libzerocash::convertBytesToVector(sha256_preimage, preimage);
+
+    std::vector<bool> expected(32*8);
+    libzerocash::convertBytesToVector(sha256_hash, expected);
+
+    // TODO: evaluate whether this should be a necessary precondition.
```
Comment (nathan-at-least) [inline tests/utilTest.cpp:308]: I don't understand this comment. The code LGTM, but the comment makes me wonder if I'm missing something.

File: tests/utilTest.cpp [hunk 19]
```diff
@@ -0,0 +1,421 @@
...
+
+    std::vector<unsigned char> expected(32);
+    libzerocash::convertBytesToBytesVector(sha256_hash, expected);
+
+    // TODO: evaluate whether this should be a necessary precondition.
+    std::vector<unsigned char> actual(32);
+    libzerocash::hashVector(preimage, actual);
+
+    BOOST_CHECK( expected == actual );
+}
```
Comment (nathan-at-least) [inline tests/utilTest.cpp:358]: LGTM. I like these tests which combine multiple conversion paths with `sha256`. These are testing the integration of conversion and hash compression, so they're kind of like integration tests. I think they're valuable, because I suspect if there's any "silent" noise like silent truncation, uninitialized bits, length extension, etc... then this layer of testing would discover this by dint of hash output inconsistencies.

File: tests/zerocashTest.cpp [hunk 20]
```diff
@@ -638,35 +640,5 @@ bool SimpleTxTest(const size_t tree_depth) {
...
-    std::cout << "\nCoinTest result => " << coinResult << std::endl;
-    std::cout << "\nMerkleTreeSimpleTest result => " << merkleSimpleResult << std::endl;
-    std::cout << "\nMintTxTest result => " << mintTxResult << std::endl;
-    std::cout << "\nPourTxTest result => " << pourTxResult << std::endl;
-    std::cout << "\nSimpleTxTest result => " << simpleTxResult << std::endl;
-    std::cout << "\nSaveAndLoadKeysFromFiles result => " << saveAndLoadKeysFromFiles << std::endl;
-
-    return !(merkleSimpleResult && addressResult &&
-        coinResult && mintTxResult && saveAndLoadKeysFromFiles
-        && pourTxResult && simpleTxResult);
```
Comment (nathan-at-least) [inline tests/zerocashTest.cpp:671]: Yay for nuking home-brew frameworks for something standard!

**Comment by nathan-at-least** [comment]
Finished code review. No blockers, and all the test code looks good.

The API, on the other hand, I find troublesome in a couple of ways. The main issue is "non-strictness": silently doing "something" if the input parameters aren't correctly sized. I believe that's error prone which is bad especially for security-sensitive code. The other is that I'm suspicious that we can't replace some chunks of this code with standard library APIs (but I may be wrong, not familiar enough yet with C++ stdlib & boost).

Off to check buildbot test results.

**Comment by nathan-at-least** [comment]
From buildbot, here are test failures I see:

```
tests/zerocashTest.cpp(532): error in "MerkleTreeSimpleTest": check (witness.at(0) == wit3) && (witness.at(1) == wit2) && (witness.at(2) == wit1) failed
```

This is a known issue already tracked in https://github.com/Electric-Coin-Company/zerocashd/issues/455 so I'm confident it's not related to these changes.

Next issue is that I don't see these new unittests being run.  I think our approach to that is to complete https://github.com/Electric-Coin-Company/zerocashd/issues/446 so I advocate merging this now and that ticket soon.

Nice work!

File: Makefile (hunk 1 above)
Comment (defuse) [inline Makefile:2]: Opened https://github.com/Electric-Coin-Company/zerocashd/issues/457

File: tests/utilTest.cpp (hunk 7 above)
Comment (defuse) [inline tests/utilTest.cpp:98]: Yeah, I think so.

File: tests/utilTest.cpp (hunk 11 above)
Comment (defuse) [inline tests/utilTest.cpp:157]: ULL is at least 8 bytes (unsigned long long, which is at least 64 bits).

File: tests/utilTest.cpp (hunk 18 above)
Comment (defuse) [inline tests/utilTest.cpp:308]: What I'm worried about is this: `hashVector` uses the size of `actual` in the loop to copy the hash, so if you _don't_ pre-allocate it before calling, you'll get an empty vector as the hash without any sort of error. That could obviously lead to security vulnerabilities.

**Comment by defuse** [comment]
Superseded by #16.

File: tests/utilTest.cpp (hunk 11 above)
Comment (daira) [inline tests/utilTest.cpp:157]: ULL used to be 128 bits on Alpha (may it rest in peace). ~~I believe that it's no longer allowed to be 128 bits in C99, but~~ I have no idea about C++.
